pandas/core/groupby/generic.py

```diff
@@ -1569,6 +1574,198 @@ def func(df):
+        normalize: bool = False,
+        sort: bool = True,
+        ascending: bool = False,
+        dropna: bool = True,
+    ) -> DataFrame | Series:
+        """
+        Return a Series or DataFrame containing counts of unique rows.
+
+        .. versionadded:: 1.4.0
+
+        Parameters
+        ----------
+        subset : list-like, optional
+            Columns to use when counting unique combinations.
+        normalize : bool, default False
+            Return proportions rather than frequencies.
+        sort : bool, default True
+            Sort by frequencies.
+        ascending : bool, default False
+            Sort in ascending order.
+        dropna : bool, default True
+            Don’t include counts of rows that contain NA values.
+
+        Returns
+        -------
+        Series or DataFrame
+            Series if as_index is True, otherwise DataFrame.
+
+        See Also
+        --------
+        Series.value_counts: Equivalent method on Series.
+        DataFrame.value_counts: Equivalent method on DataFrame.
+        SeriesGroupBy.value_counts: Equivalent method on SeriesGroupBy.
+
+        Notes
+        -----
+        If the groupby as_index is True then the returned Series will have a
+        MultiIndex with one level per input column.
+        If the groupby as_index is False then the returned DataFrame will have an
+        additional column with the value_counts.
+        By default, rows that contain any NA values are omitted from
+        the result. By default, the result will be in descending order so that the
+        first element of each group is the most frequently-occurring row.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame({
+        ...    'gender': ['male', 'male', 'female', 'male', 'female', 'male'],
+        ...    'education': ['low', 'medium', 'high', 'low', 'high', 'low'],
+        ...    'country': ['US', 'FR', 'US', 'FR', 'FR', 'FR']
+        ... })
+
+        >>> df
+            gender 	education 	country
+        0 	male 	low 	    US
+        1 	male 	medium 	    FR
+        2 	female 	high 	    US
+        3 	male 	low 	    FR
+        4 	female 	high 	    FR
+        5 	male 	low 	    FR
+
+        >>> df.groupby("gender").value_counts(normalize=True)
+        gender  education  country
+        female  high       FR         0.50
+                           US         0.50
+        male    low        FR         0.50
+                           US         0.25
+                medium     FR         0.25
+        Name: count, dtype: float64
+        """
+        RESULT_NAME = "count"
+        if self.axis == 1:
+            raise NotImplementedError(
+                "DataFrameGroupBy.value_counts only handles axis=0"
+            )
+
+        with self._group_selection_context():
+            df = self.obj
+
+            # Check for mapping, array or function rather than column grouping
+            non_column_grouping = not self.grouper.groupings[0].in_axis
```

Comment:
Just to avoid inventing new terms in the code, can we call this in_axis (remove the not), and then check `if not in_axis` below.

Reply:
Also, this and the next block of code is not sufficient, groupby can accept an iterable that combines various types of groupers. E.g.

```
df = pd.DataFrame({"A": [1, 2, 1], "B": [1, 2, 3]})
res = df.groupby(['A', [0, 1, 2]]).value_counts()
print(res)
```

This raises but should be supported.

Reply:
agree *all* of this is in the .grouper already, the key is to re-use as much code as possible to avoid inconsistent behavior.

Reply:
>  call this in_axis (remove the not), and then check if not in_axis

Done

Reply:
>  groupby can accept an iterable that combines various types of groupers

Fixed and test added (test_mixed_groupings). Please confirm that test is sufficient.